src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,2019 @@
+        asset, tenthBipsOfValue(overpayment, overpaymentFeeRate), loanScale);
+
+    Number const payment = overpayment - fee;
+
+    auto const [rawOverpaymentInterest, rawOverpaymentManagementFee] = [&]() {
+        Number const interest =
+            tenthBipsOfValue(payment, overpaymentInterestRate);
+        return detail::computeInterestAndFeeParts(interest, managementFeeRate);
+    }();
+    auto const [roundedOverpaymentInterest, roundedOverpaymentManagementFee] =
+        [&]() {
+            Number const interest =
+                roundToAsset(asset, rawOverpaymentInterest, loanScale);
+            return detail::computeInterestAndFeeParts(
+                asset, interest, managementFeeRate, loanScale);
+        }();
+
+    return detail::PaymentComponentsPlus{
+        detail::PaymentComponents{
+#if LOANCOMPLETE
+            .rawInterest = rawOverpaymentInterest,
+            .rawPrincipal = payment - rawOverpaymentInterest,
+            .rawManagementFee = 0,
+#endif
+            .trackedValueDelta = payment,
+            .trackedPrincipalDelta = payment - roundedOverpaymentInterest -
+                roundedOverpaymentManagementFee,
+            .trackedManagementFeeDelta = roundedOverpaymentManagementFee,
+            .specialCase = detail::PaymentSpecialCase::extra},
+        fee,
+        roundedOverpaymentInterest};
+}
+
+}  // namespace detail
+
+detail::LoanDeltas
+operator-(LoanState const& lhs, LoanState const& rhs)
+{
+    detail::LoanDeltas result{
+        .principalDelta = lhs.principalOutstanding - rhs.principalOutstanding,
+        .interestDueDelta = lhs.interestDue - rhs.interestDue,
+        .managementFeeDueDelta = lhs.managementFeeDue - rhs.managementFeeDue,
+    };
+
+    return result;
+}
+
+LoanState
+operator-(LoanState const& lhs, detail::LoanDeltas const& rhs)
+{
+    LoanState result{
+        .valueOutstanding = lhs.valueOutstanding - rhs.valueDelta(),
+        .principalOutstanding = lhs.principalOutstanding - rhs.principalDelta,
+        .interestDue = lhs.interestDue - rhs.interestDueDelta,
+        .managementFeeDue = lhs.managementFeeDue - rhs.managementFeeDueDelta,
+    };
+
+    return result;
+}
+
+Number
+calculateFullPaymentInterest(
+    Number const& rawPrincipalOutstanding,
+    Number const& periodicRate,
+    NetClock::time_point parentCloseTime,
+    std::uint32_t paymentInterval,
+    std::uint32_t prevPaymentDate,
+    std::uint32_t startDate,
+    TenthBips32 closeInterestRate)
+{
+    // If there is more than one payment remaining, see if enough was
+    // paid for a full payment
+    auto const accruedInterest = detail::loanAccruedInterest(
+        rawPrincipalOutstanding,
+        periodicRate,
+        parentCloseTime,
+        startDate,
+        prevPaymentDate,
+        paymentInterval);
+    XRPL_ASSERT(
+        accruedInterest >= 0,
+        "ripple::detail::computeFullPaymentInterest : valid accrued interest");
+
+    auto const prepaymentPenalty =
```

Comment:
Another small optimisation to skip multiplication:

```suggestion
    auto const prepaymentPenalty = closeInterestRate == beast::zero
        ? Number{}
        : tenthBipsOfValue(rawPrincipalOutstanding, closeInterestRate);
    
```